x-pack/plugins/security/server/authentication/providers/pki.test.ts

```diff
@@ -0,0 +1,534 @@
+    });
+
+    it('invalidates token and fails with 401 if state is present, but peer certificate is not.', async () => {
+      const request = httpServerMock.createKibanaRequest({ socket: getMockSocket() });
+      const state = { accessToken: 'token', peerCertificateFingerprint256: '2A:7A:C2:DD' };
+
+      const authenticationResult = await provider.authenticate(request, state);
+
+      expect(authenticationResult.failed()).toBe(true);
+      expect(getErrorStatusCode(authenticationResult.error)).toBe(401);
+
+      expect(mockOptions.tokens.invalidate).toHaveBeenCalledTimes(1);
+      expect(mockOptions.tokens.invalidate).toHaveBeenCalledWith({
+        accessToken: state.accessToken,
+      });
+    });
+
+    it('gets an access token in exchange to peer certificate chain and stores it in the state.', async () => {
+      const user = mockAuthenticatedUser();
+      const request = httpServerMock.createKibanaRequest({
+        headers: {},
+        socket: getMockSocket({
+          authorized: true,
+          peerCertificate: getMockPeerCertificate(['2A:7A:C2:DD', '3B:8B:D3:EE']),
+        }),
+      });
+
+      const mockScopedClusterClient = elasticsearchServiceMock.createScopedClusterClient();
+      mockScopedClusterClient.callAsCurrentUser.mockResolvedValue(user);
+      mockOptions.client.asScoped.mockReturnValue(
+        (mockScopedClusterClient as unknown) as jest.Mocked<ScopedClusterClient>
+      );
+      mockOptions.client.callAsInternalUser.mockResolvedValue({ access_token: 'access-token' });
+
+      const authenticationResult = await provider.authenticate(request);
+
+      expect(mockOptions.client.callAsInternalUser).toHaveBeenCalledTimes(1);
+      expect(mockOptions.client.callAsInternalUser).toHaveBeenCalledWith('shield.delegatePKI', {
+        body: {
+          x509_certificate_chain: [
+            'fingerprint:2A:7A:C2:DD:base64',
+            'fingerprint:3B:8B:D3:EE:base64',
+          ],
+        },
+      });
+
+      expect(mockOptions.client.asScoped).toHaveBeenCalledTimes(1);
+      expect(mockOptions.client.asScoped).toHaveBeenCalledWith({
+        headers: { authorization: `Bearer access-token` },
+      });
+      expect(mockScopedClusterClient.callAsCurrentUser).toHaveBeenCalledTimes(1);
+      expect(mockScopedClusterClient.callAsCurrentUser).toHaveBeenCalledWith('shield.authenticate');
+
+      expect(request.headers).not.toHaveProperty('authorization');
+      expect(authenticationResult.succeeded()).toBe(true);
+      expect(authenticationResult.user).toBe(user);
+      expect(authenticationResult.authHeaders).toEqual({ authorization: 'Bearer access-token' });
+      expect(authenticationResult.authResponseHeaders).toBeUndefined();
+      expect(authenticationResult.state).toEqual({
+        accessToken: 'access-token',
+        peerCertificateFingerprint256: '2A:7A:C2:DD',
+      });
+    });
+
+    it('invalidates existing token and gets a new one if fingerprints do not match.', async () => {
```

Comment:
__note:__ I was choosing between two alternatives here:
* invalidate token for the previous cert, generate new token for new cert and process the request, user won't notice anything even if new certificate represents different identity (straightforward to implement)
*  invalidate token for the previous cert, generate new token for new cert and redirect user to `overwritten_session` view that we use for similar case with SAML, but unlike SAML we may have AJAX requests with another certificate and we'll have to return error here. `400` will be quite annoying, and `401` will clear session, redirect user to the home page and may automatically re-authenticate user and that will be exactly same behavior as in the first case.

I chose the simplest (1st) option for this stage, but would like to hear what you think?

Reply:
Good! The only imaginary use case I had in mind that could potentially confuse user is when client certs are managed by some policy that automatically provisions or revokes client certificates to the user/employee browsers. But I'd expect browser to show a prompt in this case. So, let's see how it goes and re-consider in the future if needed.